### PR TITLE
Enumerables

### PR DESCRIPTION
**Pull request by uryelah**
This project is about implementing your own version of some of the Enumerable module methods in order to reinforce and apply ruby syntax.

The methods are rebuilt based on the specs at [ruby-doc.org](https://ruby-doc.org/core-2.6.5/Enumerable.html) and are the following:

- #my_each

- #my_each_with_index

- #my_select

- #my_all?

- #my_any?

- #my_none?

- #my_count

- #my_map

- #my_inject

The built Enumerable methods are inside the enumerables.rb file and some helper methods meant to decrease the complexity of the #my_inject module are located at the helpers.rb file.

### CODE REVIEW

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,93 @@
+# frozen_string_literal: true
+
+require './helpers.rb'
+
+module Enumerable
+  def my_each
+    i = 0
+    while i < length
+      yield(self[i])
```
Comment (samgaco) [inline enumerables.rb]: - [x] my_each returns an Enumerator if no block is given

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+      i += 1
+    end
+  end
+
+  def my_each_with_index
+    i = 0
+    while i < length
+      yield(self[i], i)
```
Comment (samgaco) [inline enumerables.rb]: - [x] my_each_with_index returns an enumerator if no block is given

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+      i += 1
+    end
+  end
+
+  def my_select
+    new_arr = []
+    my_each { |n| new_arr.push(n) if yield(n) }
+    new_arr
```
Comment (samgaco) [inline enumerables.rb:34]: - [x] my_select returns an enumerator if no block is given

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  end
+
+  def my_all?(pattern = nil)
+    my_each { |n| return false unless n.to_s.match(pattern) } if pattern
+    my_each { |n| return false unless yield(n) } unless pattern
```
Comment (samgaco) [inline enumerables.rb]: - [x] my_all? when no block or argument is given returns true when none of the collection members are false or nil
- [x] my_all? when a class is passed as an argument returns true if all of the collection is a member of such class
- [x] my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

File: enumerables.rb [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    true
+  end
+
+  def my_any?(pattern = nil)
+    my_each { |n| return true if n.to_s.match(pattern) } if pattern
+    my_each { |n| return true if yield(n) } unless pattern
+    false
```
Comment (samgaco) [inline enumerables.rb:56]: - [x] my_any? when no block or argument is given returns true if at least one of the collection is not false or nil
- [x] my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class

File: enumerables.rb [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+  end
+
+  def my_none?(pattern = nil)
+    my_each { |n| return false if n.to_s.match(pattern) } if pattern
+    my_each { |n| return false if yield(n) } unless pattern
+    true
```
Comment (samgaco) [inline enumerables.rb:67]: - [x] my_none? when no block or argument is given returns true only if none of the collection members is true
- [x] my_none? when a class is passed as an argument returns true if none of the collection is a member of such class
- [x] my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

File: enumerables.rb [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+      my_each { |n| new_arr << n } unless block_given?
+      new_arr = Enumerator.new(new_arr) unless block_given?
+      new_arr
+    end
+  end
+
+  def my_inject(*args)
+    memo = nil
+    if oop?(args[0])
+      initial = args[1]
```
Comment (samgaco) [inline enumerables.rb:98]: - [x] my_inject combines all elements of enum by applying a binary operation, specified by a block
- [x] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method

**Review by samgaco — CHANGES_REQUESTED** [review]
Hi @uryelah :high_brightness: 

I left you some comments below about what functionalities your methods are lacking. Please contrast your methods against the original ones [here](https://ruby-doc.org/core-2.6.3/Enumerable.html).

If you would need further clarification with some concrete examples you may **ask me in the comments** below or slack, my name is _Samuel García_ over there.

This project is a bit tricky but it will help you with getting to know ruby better, so you will be feeling more fluent moving on, therefore good spirit :muscle: 


--------------------------

* ###### STATUS OF THE PROJECT: REQUESTED CHANGES :stop_sign:

File: enumerables.rb (hunk 7 above)
Comment (uryelah) [reply enumerables.rb:98]: @samgaco 
I didn't understand this comment. It already combines the elements and uses the symbol specified when provided

Comment (samgaco) [reply enumerables.rb:98]: Hi @uryelah 

 I'll leave you some examples:

**When using it with a range gives a failure:**

```
irb(main):092:0> (1..2).inject(1){ |acc, n| acc + n}
=> 4
irb(main):093:0> (1..2).inject(2){ |acc, n| acc + n}
=> 5
irb(main):094:0> (1..2).inject(10){ |acc, n| acc + n}
=> 13
irb(main):095:0> (1..2).my_inject(10){ |acc, n| acc + n}
Traceback (most recent call last):
....
NoMethodError (undefined method `oop?' for 1..2:Range)
```

**also with an array**

```
irb(main):096:0> [1,2].inject(10){ |acc, n| acc + n}
=> 13
irb(main):097:0> [1,2].my_inject(10){ |acc, n| acc + n}
Traceback (most recent call last):
...
irb(main):096:0> [1,2].inject(10){ |acc, n| acc + n}
=> 13
irb(main):097:0> [1,2].my_inject(10){ |acc, n| acc + n}
Traceback (most recent call last):

```


**and for the second case same error:**

```
irb(main):098:0> [1,2].inject(2,:+)
=> 5
irb(main):099:0> [1,2].my_inject(2,:+)
Traceback (most recent call last):
....
NoMethodError (undefined method `oop?' for [1, 2]:Array)
```

Comment (uryelah) [reply enumerables.rb:98]: Fixed, now also works with ranges

Comment (uryelah) [reply enumerables.rb:98]: Thanks! Fixed!

Comment (uryelah) [reply enumerables.rb:98]: I did not get the 
`NoMethodError (undefined method `oop?' for 1..2:Range)`
error.
oop? is a class defined in the helpers.rb and it's being required at the top of the enumerables.rb file so it should be defined

**Review by maelfosso — APPROVED** [review]
# Awesome :clap: The project is approved :smile: 
You can merge the branch now !!!
Good luck with the next project.

**Mael FOSSO**
[Github](https://github.com) | [Twitter](https://twitter.com) | [LinkedIn](https://www.linkedin.com/in/mael-elvis-fosso-650b6346/)

**Pull request merged**